[WIP] User list section/filters

srdeotarse opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
This WIP PR migrates earlier struture of ``RightColumnView`` in ``views.py`` to structure similar to ``LeftColumnView`` with toggling between ``StreamsView`` and ``TopicsView``.
This PR mainly intends to add new ``RecipientsView`` that displays recipients of the current narrow(``All_messages``, ``PMs``, ``Streams``, ``Topics`` etc) in the right column.
<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
This fixes #575
<!-- Add a link to a discussion on chat.zulip.org, if relevant -->
CZO - https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/User.20list.20sections.2Ffilters.20.23T575
**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [ ] Manually
- [ ] Existing tests (adapted, if necessary)
- [ ] New tests added (for any new behavior)
- [ ] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->
**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->
This WIP PR only includes transition from earlier structure of ``RightColumnView`` to new structure based on ``LeftColumnView`` structure. It includes modifications to ``UserView``
**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->
This PR is motivated by PR #459
**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->

neiljp commented:
@srdeotarse We discussed this in the stream, so I'm marking it as awaiting update on the assumption that we've clarified how to move forward with this. Let me know if you need more feedback/review.

srdeotarse committed:
refactor: views: Spilt users_view function in RightColumnView.

Split the function into two functions -
1) build_user_view - prepares the Users button list based on users
input and returns the user button list.
2) users_view - calls build_user_view to use user button list
to prepare the UsersView.

This refactor will help for adding Recipients list in addition to
current Users list and have toggling feature between then similar to
streams and topics view.

srdeotarse committed:
refactor: ui/views: Shift 'Users' urwid.LineBox from to users_view.

Shifted from rigth_column_view in ui.py to users_view in
RightColumnView in views.py.

srdeotarse committed:
model: Add get_recipients_of_current_narrow function.

Based on current narrow, the function return recipients list.
All_messages and streams narrow is handled.
PMs_all, PMs(1 to 1) and Groups narrow remaining.

srdeotarse committed:
keys/hotkeys: Add TOGGLE_RECIPIENTS_VIEW shortcut key.

srdeotarse committed:
buttons: Add toggling from UsersView to RecipientsView.

srdeotarse committed:
refactor: views/model: Shift update_user_list and build_user_view.

Changed RightColumnView to urwid.Pile and UsersView to urwid.Frame
similar to LeftColumnView and StreamsView/TopicsView.
Shifted update_user_list and build_user_view functions from
RightColumnView to UsersView.
Change view.users_view to view.user_w in _start_presence_updates
function in model.

srdeotarse committed:
views: Add recipients_view function in RightColumnView.

Also added update_recipients_view, show_recipients_view and
show_user_view functions.

neiljp reviewed (commented):
@srdeotarse Thanks for looking at this! This is so large, and promises to be more commits, since it's really a big tidy/refactor first, followed by the feature itself :)

I seem to have written an essay of notes to help with the preliminary refactor ;) Whenever adjusting working code, it's useful to break it into almost too many independent working commits, as it tends to make each commit smaller and easier to read, and so take less energy and time to review. `git` can tell me if code has just been moved in a commit, but when it's interspersed with other changes then it's more difficult to validate.

While you could adapt tests for your giant code move in one go, it would likely be a lot of changes there too, which are also then difficult to associate with a given code change without breaking into smaller commits.

I almost suggested adding tests for the new model method, but like I said there, I'm not sure if you want to use it to provide the appropriate list/set of user ids only instead. That may be more obvious as you make the PR slightly more functional.

While you hopefully make sense of the refactoring splitting, for the functionality of the recipients part, there are various elements to address:
- toggling-back from recipients (we discussed); currently the list is only updated to the filtered view on toggling, so it would be useful to know it can be explored/tested manually by toggling back and forth in different narrows :)
- updating reliably from regular refreshes; this currently does an all-users view, since the user list assumes on updating, that the user list is the list to display. Maybe handle the filtering in the UsersView?
- updating on changing narrow; this is a similar problem to the 'highlight button representing active narrow' PR
